File: advanced/tier-2/python/run_case.py
"""Periodic conservative advection–diffusion: upwind versus Lax–Wendroff."""
from pathlib import Path
import argparse,json,csv
from time import perf_counter
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt


def solve(n,speed,alpha,length,time,scheme):
    if n<8 or alpha<0 or length<=0 or time<=0 or scheme not in ('upwind','lax_wendroff'):
        raise ValueError('Invalid transport case.')
    dx=length/n; x=(np.arange(n)+.5)*dx
    # Sufficient monotonic upwind bound; LW also obeys C²+2D<=1.
    rate=abs(speed)/dx+2*alpha/dx**2
    steps=max(1,int(np.ceil(time*rate/.8))); dt=time/steps
    C=speed*dt/dx; D=alpha*dt/dx**2
    if scheme=='lax_wendroff' and C*C+2*D>1+1e-12: raise ValueError('Unstable step.')
    u=1+.5*np.sin(2*np.pi*x/length); initial_mass=float(dx*u.sum())
    timer=perf_counter()
    for _ in range(steps):
        right=np.roll(u,-1);left=np.roll(u,1)
        if scheme=='upwind':
            flux=speed*(u if speed>=0 else right)-alpha*(right-u)/dx
        else:
            flux=speed*.5*(u+right)-.5*speed*C*(right-u)-alpha*(right-u)/dx
        u-=dt/dx*(flux-np.roll(flux,1))
    elapsed=perf_counter()-timer
    exact=1+.5*np.exp(-alpha*(2*np.pi/length)**2*time)*np.sin(2*np.pi*(x-speed*time)/length)
    return x,u,exact,dict(cells=n,scheme=scheme,dt=dt,steps=steps,courant=C,diffusion_number=D,l2_error=float(np.sqrt(dx*np.sum((u-exact)**2))),mass_drift=float(dx*u.sum()-initial_mass),integration_seconds=elapsed)


def main():
    p=argparse.ArgumentParser();base=Path(__file__).resolve().parents[1]
    p.add_argument('--case',type=Path,default=base/'cases'/'transport.json');p.add_argument('--output',type=Path,default=base/'results'/'python');args=p.parse_args()
    c=json.loads(args.case.read_text());args.output.mkdir(parents=True,exist_ok=True)
    rows=[];fig,ax=plt.subplots()
    for scheme in ('upwind','lax_wendroff'):
        for n in c['cell_counts']:
            x,u,exact,m=solve(n,c['speed'],c['diffusivity'],c['length'],c['final_time'],scheme);rows.append(m)
            np.savetxt(args.output/f'{scheme}_{n}.csv',np.c_[x,u,exact],delimiter=',',header='x,numerical,analytical',comments='')
        ax.plot(x,u,label=scheme)
    ax.plot(x,exact,'k--',label='Analytical');ax.set(xlabel='x (m)',ylabel='Scalar',title='Periodic advection–diffusion');ax.legend();fig.tight_layout();fig.savefig(args.output/'solution.png',dpi=150);plt.close(fig)
    with (args.output/'convergence.csv').open('w',newline='') as f:
        w=csv.DictWriter(f,fieldnames=list(rows[0]));w.writeheader();w.writerows(rows)
    (args.output/'run.json').write_text(json.dumps(dict(case=c,metrics=rows),indent=2))
    print(json.dumps(rows,indent=2))
if __name__=='__main__':main()

File: advanced/tier-2/tier-2/matlab/run_case.m
function run_case()
% Periodic conservative advection-diffusion; same case as Python.
base=fileparts(fileparts(mfilename('fullpath')));
c=jsondecode(fileread(fullfile(base,'cases','transport.json')));
out=fullfile(base,'results','matlab'); if ~exist(out,'dir'), mkdir(out); end
schemes={'upwind','lax_wendroff'}; rows=struct([]); q=0;
f=figure('Visible','off'); hold on;
for s=1:2
    for ni=1:numel(c.cell_counts)
        n=c.cell_counts(ni); dx=c.length/n; x=((0:n-1)'+.5)*dx;
        assert(n>=8 && c.diffusivity>=0 && c.final_time>0 && c.length>0);
        rate=abs(c.speed)/dx+2*c.diffusivity/dx^2;
        steps=max(1,ceil(c.final_time*rate/.8)); dt=c.final_time/steps;
        C=c.speed*dt/dx; D=c.diffusivity*dt/dx^2;
        assert(C^2+2*D<=1+1e-12);
        u=1+.5*sin(2*pi*x/c.length); mass=dx*sum(u); timer=tic;
        for j=1:steps
            right=circshift(u,-1);
            if s==1
                if c.speed>=0, upstream=u; else, upstream=right; end
                flux=c.speed*upstream-c.diffusivity*(right-u)/dx;
            else
                flux=c.speed*.5*(u+right)-.5*c.speed*C*(right-u)-c.diffusivity*(right-u)/dx;
            end
            u=u-dt/dx*(flux-circshift(flux,1));
        end
        elapsed=toc(timer);
        exact=1+.5*exp(-c.diffusivity*(2*pi/c.length)^2*c.final_time)*sin(2*pi*(x-c.speed*c.final_time)/c.length);
        q=q+1; rows(q)=struct('cells',n,'scheme',schemes{s},'dt',dt,'steps',steps,'courant',C,'diffusion_number',D,...
            'l2_error',sqrt(dx*sum((u-exact).^2)),'mass_drift',dx*sum(u)-mass,'integration_seconds',elapsed);
        writetable(table(x,u,exact,'VariableNames',{'x','numerical','analytical'}),fullfile(out,sprintf('%s_%d.csv',schemes{s},n)));
    end
    plot(x,u,'DisplayName',schemes{s});
end
plot(x,exact,'k--','DisplayName','Analytical');legend('show','Interpreter','none');
xlabel('x (m)');ylabel('Scalar');title('Periodic advection-diffusion');
print(f,fullfile(out,'solution.png'),'-dpng','-r150');close(f);
writetable(struct2table(rows),fullfile(out,'convergence.csv'));save(fullfile(out,'run.mat'),'c','rows');disp(struct2table(rows));
end
